aziot/src/internal/check/checks/identity_certificate_expiry.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use anyhow::Result;
+use serde::Serialize;
+
+use crate::internal::check::util::CertificateValidityExt;
+use crate::internal::check::{CheckResult, Checker, CheckerCache, CheckerMeta, CheckerShared};
+use crate::internal::common::CertificateValidity;
+
+#[derive(Serialize, Default)]
+pub struct IdentityCertificateExpiry {
+    provisioning_mode: Option<&'static str>,
+    certificate_info: Option<CertificateValidity>,
+}
+
+#[async_trait::async_trait]
+impl Checker for IdentityCertificateExpiry {
+    fn meta(&self) -> CheckerMeta {
+        CheckerMeta {
+            id: "identity-certificate-expiry",
+            description: "production readiness: identity certificates expiry",
+        }
+    }
+
+    async fn execute(&mut self, shared: &CheckerShared, cache: &mut CheckerCache) -> CheckResult {
+        self.inner_execute(shared, cache)
+            .await
+            .unwrap_or_else(CheckResult::Failed)
+    }
+}
+
+impl IdentityCertificateExpiry {
+    async fn inner_execute(
+        &mut self,
+        _shared: &CheckerShared,
+        cache: &mut CheckerCache,
+    ) -> Result<CheckResult> {
+        use aziot_identityd_config::{DpsAttestationMethod, ManualAuthMethod, ProvisioningType};
+
+        let provisioning = &unwrap_or_skip!(&cache.cfg.identityd)
+            .provisioning
+            .provisioning
+            .clone();
+
+        match provisioning {
+            ProvisioningType::Dps {
+                attestation: DpsAttestationMethod::X509 { identity_cert, .. },
+                ..
+            } => {
+                self.provisioning_mode = Some("dps-x509");
+                let cert_info = CertificateValidity::new(
+                    unwrap_or_skip!(cache.cert_path(&identity_cert))?,
+                    "DPS identity certificate",
+                    &identity_cert,
+                )
+                .await?;
+                self.certificate_info = Some(cert_info.clone());
+                return cert_info.to_check_result();
+            }
+            ProvisioningType::Manual {
+                authentication: ManualAuthMethod::X509 { identity_cert, .. },
+                ..
+            } => {
+                self.provisioning_mode = Some("manual-x509");
+                let cert_info = CertificateValidity::new(
+                    unwrap_or_skip!(cache.cert_path(&identity_cert))?,
```

Comment:
None of the `unwrap_or_skip!(cache.cert_path(cert_id))?`s are going to work if `cert_id` is a dynamically issued cert instead of a preloaded one. In that case `cert_path` returns `Some(Err)` instead of `None`

Reply:
I'm not sure I follow. Unless I'm missing something, the underlying `aziot_certd_config::util::get_path` function just hashes the cert_id if it's not preloaded, so it should still succeed.

Reply:
You're right. `get_path()` and `cache.cert_path()` will suceed. It's `CertificateValidity::new()` that will fail. So the point still stands.

Reply:
Wait, why? Wouldn't it just read the cert from disk?
Or am I misunderstanding how dynamically generated certs work?

Reply:
The point of being dynamically generated is that the cert *isn't* already on disk. It's issued dynamically by CS when someone invokes its rest API.

Reply:
Eg https://github.com/Azure/iot-identity-service/blob/7e6c062988f26299fc4695d0df55a49e48b839cf/identity/aziot-identityd/src/lib.rs#L689

Reply:
...okay, but if that method was previously called, then the cert would exist on disk, right?
Or is a brand new one generated on every run?

Reply:
>...okay, but if that method was previously called, then the cert would exist on disk, right?

If IS hasn't gotten around to provisioning yet for any reason, the device ID cert hasn't necessarily been created. You shouldn't assume that the user is running `aziot check` after IS has finished provisioning. If that was the assumption, we wouldn't need upstream connectivity checks.

Reply:
Okay, sure, but that doesn't answer my question about what the check should do if the cert is dynamically issued?

I understand things correctly, the current implementation will return a `CheckResult::Failed(<could not read file>)`, which doesn't seem totally unreasonable. One alternative could be to have a special case that checks if the cert is dynamically issued, and if the file doesn't exist in that case, report some alternative error (i.e: `no dynamically issued identity certificate has been issued`).

Any other ideas?

Reply:
Treat it as as success or skipped. (That's what I meant in the first comment when I said it ought to behave like `None`.)

The check is primarily there to catch bad certs that the user created. Even if the dynamic configuration is broken for any reason (say the EST server always issued expired certs), there's no reason to pre-emptively believe a dynamically issued cert will be bad just because it hasn't been created yet. An error here is non-actionable by the user if the cert doesn't exist yet.

Reply:
On that note, if the cert file isn't found, you should assert that the cert *is* configured to be dynamically issued according to the CS config, ie its ID is in `.cert_issuance.certs.keys()`.

Reply:
Sounds good. I'll push the necessary changes tomorrow morning.

Reply:
Pushed up a fix. Simplified the code in the process, since it turned out my `cert_path` method was only called from this one check anyways.